lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py

```diff
@@ -0,0 +1,376 @@
+from lightkube.utils.quantity import equals_canonically
+from ops.charm import CharmBase
+from ops.framework import BoundEvent, EventBase, EventSource, Object, ObjectEvents
+
+logger = logging.getLogger(__name__)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "2a6066f701444e8db44ba2f6af28da90"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+class ResourceSpecDict(TypedDict, total=False):
+    """A dict representing a K8s resource limit.
+
+    See:
+    - https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+    - https://gtsystem.github.io/lightkube-models/1.23/models/core_v1/#resourcerequirements
+    """
+
+    cpu: Optional[str]
+    memory: Optional[str]
+
+
+def sanitize_resource_spec_dict(dct: Optional[ResourceSpecDict]) -> Optional[ResourceSpecDict]:
+    """Make sure only recognized keys are present, and drop all keys whose value is empty."""
+    if not dct:
+        return dct
+
+    d = dct.copy()
+    if any(k not in ResourceSpecDict.__annotations__.keys() for k in d.keys()):
+        raise ValueError(
+            "Invalid resource limits dict: must have keys {}".format(
+                ", ".join(ResourceSpecDict.__annotations__.keys())
+            )
+        )
+    for k, v in dct.items():
+        if v == "" or v is None:
+            # Remove key, otherwise it will be serialized as 0 and pod won't be scheduled.
+            del d[k]  # type: ignore
+
+    # fractional byte value "858993459200m" (0.8Gi) is invalid, must be an integer
+    # FIXME: round up memory two whole bytes
+    return d
+
+
+class K8sResourcePatchFailedEvent(EventBase):
+    """Emitted when patching fails."""
+
+    def __init__(self, handle, message=None):
+        super().__init__(handle)
+        self.message = message
+
+    def snapshot(self) -> Dict:
+        """Save grafana source information."""
+        return {"message": self.message}
+
+    def restore(self, snapshot):
+        """Restore grafana source information."""
+        self.message = snapshot["message"]
+
+
+class K8sResourcePatchEvents(ObjectEvents):
+    """Events raised by :class:`K8sResourcePatchEvents`."""
+
+    patch_failed = EventSource(K8sResourcePatchFailedEvent)
+
+
+class ContainerNotFoundError(ValueError):
+    """Raised when a given container does not exist in the list of containers."""
+
+
+class KubernetesComputeResourcesPatch(Object):
+    """A utility for patching the Kubernetes compute resources set up by Juju."""
+
+    on = K8sResourcePatchEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        container_name: str,
+        *,
+        limits: Optional[ResourceSpecDict],
+        requests: Optional[ResourceSpecDict],
+        refresh_event: Optional[Union[BoundEvent, List[BoundEvent]]] = None,
```

Comment:
About `refresh_event`... 

Why don't we rename it as `refresh_events` that receives always a `list` no matter if it's one or more than one events?

I think this way the class contract is clearer.... besides the type hinting is cleared too